### PR TITLE
wsdd: init at 0.6.2

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Add wsdd package. This package implements a Web Service Discovery host daemon. This enables (Samba) hosts, like your local NAS device, to be found by Web Service Discovery Clients like Windows.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/wsdd/default.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    sha256 = "0444xh1r5wd0zfch1hg1f9s4cw68srrm87hqx16qvlgx6jmz5j0p";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  buildInputs = [ python3 ];
+
+  installPhase = ''
+    install -Dm0444 src/wsdd.py $out/bin/wsdd
+    chmod +x $out/bin/wsdd
```
Comment (SuperSandro2000) [inline pkgs/servers/wsdd/default.nix:20]:
```suggestion
    install -Dm0755 src/wsdd.py $out/bin/wsdd
```

File: pkgs/servers/wsdd/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  '';
+
+  meta = with stdenv.lib; {
+    homepage    = "https://github.com/christgau/wsdd";
+    description = "A Web Service Discovery (WSD) host daemon for SMB/Samba.";
+    maintainers = with maintainers; [ izorkin ];
+    license     = licenses.mit;
+    platforms   = platforms.all;
+  };
+
```
Comment (SuperSandro2000) [inline pkgs/servers/wsdd/default.nix:31]:
```suggestion
    homepage = "https://github.com/christgau/wsdd";
    description = "A Web Service Discovery (WSD) host daemon for SMB/Samba";
    maintainers = with maintainers; [ izorkin ];
    license = licenses.mit;
    platforms = platforms.all;
  };
```

File: pkgs/servers/wsdd/default.nix (hunk 1 above)
Comment (Izorkin) [reply pkgs/servers/wsdd/default.nix]: Thanks, fixed.

File: pkgs/servers/wsdd/default.nix (hunk 2 above)
Comment (Izorkin) [reply pkgs/servers/wsdd/default.nix]: Thanks, fixed.

**Comment by Izorkin** [comment]
cc @flokli @aanderse @Mic92

**Comment by flokli** [comment]
Can this be made to run with DynamicUser?

**Comment by Izorkin** [comment]
With `DynamicUser = true;` worked, updated.

**Comment by flokli** [comment]
How does this relate so samba's `wins`?

Can this trivially be included in `nixos/tests/samba.nix`, and is there a client implementation, so this can be tested?

**Comment by Izorkin** [comment]
Used this information - https://linux-blog.anracom.com/2020/05/24/samba-4-shares-wsdd-and-windows-10-how-to-list-linux-samba-servers-in-the-win-10-explorer/
I running wsl on my home server - samba server is displayed in in the Explorer. With runnig wins server - in the Explorer not displayed.

**Comment by flokli** [comment]
Look at https://www.mankier.com/1/wsdd#Options-Client/Discovery_Mode_Options - we should enable this on the server, and check with `wssd`'s discovery mode that the server can be seen.

**Comment by Izorkin** [comment]
Working without discovery - https://www.mankier.com/1/wsdd#Example_Usage-Set_the_Workgroup_according_to_smb.conf,_be_verbose,_run_with_dropped_privileges,_and_change_the_root_directory_to_an_(empty)_directory

Options `--discovery` need to search others WSD hosts/servers.

**Comment by Izorkin** [comment]
Example configuration:
```
  services.samba-wsdd = {
    enable = true;
    interface = "eno1";
    workgroup = "WORKGROUP";
    hostname = "FILESERVER";
    extraOptions = [ "--verbose" "--ipv4only" ];
  };

  services.samba = {
    enable = true;
    enableNmbd = true;
    enableWinbindd = false;
    nsswins = false;
    package = pkgs.samba4Minimal;
    securityType = "user";
    extraConfig = ''
      workgroup = WORKGROUP
      netbios name = FILESERVER
...
```

**Comment by Izorkin** [comment]
https://devanswers.co/discover-ubuntu-machines-samba-shares-windows-10-network/ - simpe WSDD configuration

**Comment by flokli** [comment]
Sorry, not sure I understand.

Please extend `nixos/tests/samba.nix` with a `wssd` configuration, so there's a working example in nixpkgs, that can be tested in an automated fashion.

Above example does not work, and `pkgs.samba4Minimal` doesn't even exist in nixpkgs, nor is there any described way on how to verify this, without booting a Windows machine and checking there. 

I tried extending the test, by adding a second server with wsdd running, and using `wsdd --discovery` on the client, hoping I could see the server, but it didn't work, even when disabling the firewall.

Maybe you can pick up from here, here's my diff:

```
diff --git a/nixos/tests/samba.nix b/nixos/tests/samba.nix
index 142269752b3..6aa43bc7a3f 100644
--- a/nixos/tests/samba.nix
+++ b/nixos/tests/samba.nix
@@ -5,34 +5,52 @@ import ./make-test-python.nix ({ pkgs, ... }:
 
   meta.maintainers = [ pkgs.lib.maintainers.eelco ];
 
-  nodes =
-    { client =
-        { pkgs, ... }:
-        { fileSystems = pkgs.lib.mkVMOverride
-            { "/public" = {
-                fsType = "cifs";
-                device = "//server/public";
-                options = [ "guest" ];
-              };
-            };
+  nodes = {
+    client = { pkgs, ... }: {
+      fileSystems = pkgs.lib.mkVMOverride {
+        "/public" = {
+          fsType = "cifs";
+          device = "//server/public";
+          options = [ "guest" ];
         };
+      };
+      networking.firewall.allowedTCPPorts = [ 5357 ];
+      networking.firewall.allowedUDPPorts = [ 3702 ];
+      networking.firewall.enable = false;
+    };
 
-      server =
-        { ... }:
-        { services.samba.enable = true;
-          services.samba.shares.public =
-            { path = "/public";
-              "read only" = true;
-              browseable = "yes";
-              "guest ok" = "yes";
-              comment = "Public samba share.";
-            };
-          networking.firewall.allowedTCPPorts = [ 139 445 ];
-          networking.firewall.allowedUDPPorts = [ 137 138 ];
-        };
+    server = { ... }: {
+      services.samba.enable = true;
+      services.samba.shares.public = {
+        path = "/public";
+        "read only" = true;
+        browseable = "yes";
+        "guest ok" = "yes";
+        comment = "Public samba share.";
+      };
+      networking.firewall.allowedTCPPorts = [ 139 445 ];
+      networking.firewall.allowedUDPPorts = [ 137 138 ];
     };
 
-  # client# [    4.542997] mount[777]: sh: systemd-ask-password: command not found
+    server_wsdd = { ... }: {
+      services.samba = {
+        enable = true;
+        enableNmbd = true;
+        enableWinbindd = false;
+        nsswins = false;
+      };
+      services.samba-wsdd = {
+        enable = true;
+        interface = "eth0";
+        workgroup = "WORKGROUP";
+        hostname = "SERVER-WSDD";
+        extraOptions = [ "--verbose" ];
+      };
+      networking.firewall.allowedTCPPorts = [ 139 445 5357 ];
+      networking.firewall.allowedUDPPorts = [ 137 138 3702 ];
+      networking.firewall.enable = false;
+    };
+  };
 
   testScript =
     ''
@@ -43,5 +61,16 @@ import ./make-test-python.nix ({ pkgs, ... }:
       client.start()
       client.wait_for_unit("remote-fs.target")
       client.succeed("[[ $(cat /public/foo) = bar ]]")
+
+      server_wsdd.start()
+      server_wsdd.wait_for_unit("samba-wsdd")
+
+      client.succeed(
+          "systemd-run -u wsdd-discovery -- ${pkgs.wsdd}/bin/wsdd --verbose --discovery --preserve-case -l /run/wsdd.sock >&2"
+      )
+
+      client.wait_until_succeeds(
+          "echo list | ${pkgs.libressl.nc}/bin/nc -U /run/wsdd.sock | grep -i SERVER-WSDD"
+      )
     '';
 })
```

**Comment by Izorkin** [comment]
Working if runnig in test VM
wsdd --interface eth0 --verbose --discovery
```
2020-11-23 10:12:38,157:wsdd INFO(pid 2046): using pre-defined UUID fd3a88ed-...-...
2020-11-23 10:12:38,157:wsdd WARNING(pid 2046): Discovery enabled but no listen option provided. Falling back to port 5359
2020-11-23 10:12:38,159:wsdd INFO(pid 2046): joined multicast group ('239.255.255.250', 3702) on 192. ... .221%eth0
2020-11-23 10:12:40,248:wsdd INFO(pid 2046): joined multicast group ('ff02::c', 3702, 22364, 2) on fe80::...:d2d5%eth0
2020-11-23 10:12:43,254:wsdd INFO(pid 2046): 192. ... .221:44651(eth0) - - "ProbeMatches urn:uuid:45e06928-...-... UDP" - -
2020-11-23 10:12:43,255:wsdd INFO(pid 2046): 192. ... .221:44651(eth0) - - "ResolveMatches urn:uuid:47aae12f-...-... UDP" - -
2020-11-23 10:12:43,261:wsdd INFO(pid 2046): discovered TECH in Workgroup:WORKGROUP on 192. ... .221%eth0
```
With nixos/tests i can't do the test.

**Comment by Izorkin** [comment]
With runnin command `nc -U /run/wsdd/wsdd.sock` not correct working discovering function

**Comment by Izorkin** [comment]
Updatet PR and added test WSD Discovery.

**Comment by Izorkin** [comment]
@GrahamcOfBorg test samba

File: nixos/tests/samba.nix [hunk 3]
```diff
@@ -5,34 +5,52 @@ import ./make-test-python.nix ({ pkgs, ... }:
...
-            };
+  nodes = {
+    client = { pkgs, ... }: {
+      services.samba-wsdd = {
+        enable = true;
+        interface = "eth1";
+        workgroup = "WORKGROUP";
+        hostname = "CLIENT-WSDD";
+        discovery = true;
+        extraOptions = [ "--verbose" "--no-host" ];
```
Comment (flokli) [inline nixos/tests/samba.nix:16]: Let's drop this line, it should still work.

File: nixos/tests/samba.nix [hunk 4]
```diff
@@ -5,34 +5,52 @@ import ./make-test-python.nix ({ pkgs, ... }:
...
+      };
+      fileSystems = pkgs.lib.mkVMOverride {
+        "/public" = {
+          fsType = "cifs";
+          device = "//server/public";
+          options = [ "guest" ];
         };
+      };
+      networking.firewall.allowedTCPPorts = [ 5357 ];
+      networking.firewall.allowedUDPPorts = [ 3702 ];
```
Comment (flokli) [inline nixos/tests/samba.nix:26]: We should mention these ports are required for wsdd discovery only. Or maybe entirely put this into another `client_discovery` VM?

File: nixos/tests/samba.nix [hunk 5]
```diff
@@ -5,34 +5,52 @@ import ./make-test-python.nix ({ pkgs, ... }:
...
     };
 
-  # client# [    4.542997] mount[777]: sh: systemd-ask-password: command not found
+    server_wsdd = { ... }: {
+      services.samba-wsdd = {
+        enable = true;
+        interface = "eth1";
+        workgroup = "WORKGROUP";
+        hostname = "SERVER-WSDD";
+        extraOptions = [ "--verbose" ];
```
Comment (flokli) [inline nixos/tests/samba.nix:48]: We can drop this line

File: nixos/tests/samba.nix [hunk 6]
```diff
@@ -5,34 +5,52 @@ import ./make-test-python.nix ({ pkgs, ... }:
...
+      };
+      networking.firewall.allowedTCPPorts = [ 5357 ];
+      networking.firewall.allowedUDPPorts = [ 3702 ];
```
Comment (flokli) [inline nixos/tests/samba.nix:51]: Same here, comment this is only necessary for discovery.

File: pkgs/servers/wsdd/default.nix [hunk 7]
```diff
@@ -20,6 +20,10 @@ stdenv.mkDerivation rec {
     wrapProgram $out/bin/wsdd --prefix PYTHONPATH : "$PYTHONPATH"
   '';
 
+  passthru = {
+    tests.samba = nixosTests.samba;
```
Comment (flokli) [inline pkgs/servers/wsdd/default.nix:24]: :+1:

File: nixos/tests/samba.nix (hunk 4 above)
Comment (Izorkin) [reply nixos/tests/samba.nix:26]: Can I transfer it to a separate test - samba-wsdd?

File: nixos/tests/samba.nix (hunk 3 above)
Comment (Izorkin) [reply nixos/tests/samba.nix:16]: If this drop line - the test freezing. Needed options `"--no-host"`

**Comment by Izorkin** [comment]
@flokli updated PR.

**Review by flokli — APPROVED** [review]

**Comment by Izorkin** [comment]
@GrahamcOfBorg test samba-wsdd

**Review by Mic92 — APPROVED** [review]
Untested, but LGTM.

**Pull request merged**